h/streamer/messages.py

```diff
@@ -103,40 +111,36 @@ def handle_annotation_event(message, sockets, registry, session):
     # Create a generator which has the first socket back again
     matching_sockets = chain((first_socket,), matching_sockets)
 
-    nipsa_service = NipsaService(session)
-    user_nipsad = nipsa_service.is_flagged(annotation.userid)
-
     authority = registry.settings.get("h.authority", "localhost")
-    group_service = GroupfinderService(session, authority)
-    user_service = UserService(authority, session)
-    formatters = [AnnotationUserInfoFormatter(session, user_service)]
+    base_url = registry.settings.get("h.app_url", "http://localhost:5000")
+
+    resource = AnnotationNotificationContext(
+        annotation,
+        group_service=GroupfinderService(session, authority),
+        links_service=LinksService(base_url, registry),
+    )
+    read_principals = principals_allowed_by_permission(resource, "read")
+    reply = _generate_annotation_event(session, authority, message, resource)
+
+    annotator_nipsad = NipsaService(session).is_flagged(annotation.userid)
 
     for socket in matching_sockets:
-        reply = _generate_annotation_event(
-            message,
-            socket,
-            annotation,
-            user_nipsad,
-            group_service,
-            formatters,
-            registry,
-        )
-        if reply is None:
+        # Don't send notifications back to the person who sent them
+        if message["src_client_id"] == socket.client_id:
             continue
-        socket.send_json(reply)
 
+        # Only send NIPSA'd annotations to the author
+        if annotator_nipsad and socket.authenticated_userid != annotation.userid:
+            continue
 
-def handle_user_event(message, sockets, registry, session):
-    for socket in sockets:
-        reply = _generate_user_event(message, socket)
-        if reply is None:
+        # Check whether client is authorized to read this annotation.
+        if not set(read_principals).intersection(socket.effective_principals):
             continue
+
```

Comment:
It looks like the per-socket loop is now cut down to the minimum, only the work that actually varies for each socket